File: 백준/Gold/10564. 팔굽혀펴기/팔굽혀펴기.cc
#include <bits/stdc++.h>
using namespace std;
#define fastio                        \
    ios_base::sync_with_stdio(false); \
    cin.tie(NULL)
#define ll long long

int arr[15];
int cache[5005][5005];
int n, m;
int caching(int begin, int sum)
{
    if (sum == n)
        return 0;
    if (sum > n)
        return -987654321;
    int &ret = cache[begin][sum];
    if (ret != -1)
        return ret;
    ret = -987654321;
    for (int i = 0; i < m; i++)
    {
        // cout << begin << ' ' << arr[i] << '\n';
        ret = max(ret, caching(begin + arr[i], sum + begin + arr[i]) + arr[i]);
    }
    return ret;
}
void solve()
{
    cin >> n >> m;
    for (int i = 0; i < m; i++)
    {
        cin >> arr[i];
    }

    memset(cache, -1, sizeof(cache));
    int ret = caching(0, 0);
    if (ret < 0)
        ret = -1;
    cout << ret << '\n';
}

int main()
{
    fastio;
    int t;
    cin >> t;
    while (t--)
    {
        solve();
    }

    return 0;
}

File: 백준/Gold/10564. 팔굽혀펴기/팔굽혀펴기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.Buffer;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int[] arr;
    static int[][] cache;
    static int n, m;

    static int caching(int sum, int turn) {
        if (sum == 0)
            return 0;
        if (cache[sum][turn] != -1)
            return cache[sum][turn];
        int ret = Integer.MIN_VALUE;
        for (int i = 0; i < m; i++) {
            if (sum - turn * arr[i] >= 0)
                ret = Math.max(ret, caching(sum - turn * arr[i], turn + 1) + arr[i]);
        }
        cache[sum][turn] = ret;
        return ret;
    }

    static void solve(BufferedReader br) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new int[m];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < m; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        cache = new int[n + 5][105];
        for (int i = 0; i <= n; i++) {
            Arrays.fill(cache[i], -1);
        }
        int ret = caching(n, 1);
        if (ret < 0) {
            ret = -1;
        }
        System.out.println(ret);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            solve(br);
        }
    }
}
